File: elixir/isogram/isogram.exs
defmodule Isogram do
  @doc """
  Determines if a word or sentence is an isogram
  """
  @spec isogram?(String.t) :: boolean
  def isogram?(sentence) do
    sentence
    |> sanitize
    |> String.graphemes
    |> Enum.uniq
    |> Enum.join
    |> unchanged?(sanitize(sentence))
  end

  defp sanitize(sentence) do
    sentence
    |> String.replace(~r/[- ]/, "")
  end

  defp unchanged?(deduped, original) do
    deduped == original
  end
end

File: javascript/isogram/isogram.js
function Isogram(word) {
  const sanitized = sanitize(word.toLowerCase())
  const deduped = dedup(sanitized)
  this.is = deduped.length == sanitized.length
}

Isogram.prototype.isIsogram = function () {
  return this.is
}

function sanitize(word) {
  return word.replace(/[- ]/g, "")
}

function dedup(word) {
  return Array.from(new Set(word.split("")))
}

module.exports = Isogram
